### PR TITLE
Add multi-ncclcomm and hierarchical ncclallreduce support.

### LINKED ISSUE #17456: `fused_all_reduce_op` hangs in trainning.

### PR DESCRIPTION
**Pull request by gongweibao**
1. Add multi nccl comm.
2. Add 2d nccl allreduce.
3. Fix https://github.com/PaddlePaddle/Paddle/issues/17456
4. Fine tune ncclid generation.

image

### CODE REVIEW

File: paddle/fluid/framework/details/all_reduce_op_handle.h [hunk 1]
```diff
@@ -46,13 +49,16 @@ class AllReduceOpHandle : public OpHandleBase {
...
  protected:
   void RunImpl() override;
-
   std::vector<Scope *> local_scopes_;
+
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+#else
   std::vector<platform::Place> places_;
+#endif
+
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/all_reduce_op_handle.h:59]: Why the `if then` is empty here?

File: paddle/fluid/framework/details/all_reduce_op_handle.h [hunk 2]
```diff
@@ -21,20 +21,23 @@
...
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+class AllReduceOpHandle : public NCCLOpHandleBase {
+ public:
   AllReduceOpHandle(ir::Node *node, const std::vector<Scope *> &local_scopes,
                     const std::vector<platform::Place> &places,
-                    const platform::NCCLContextMap *ctxs);
+                    const platform::MultiNCCLContextMap *ctxs);
 #else
+class AllReduceOpHandle : public CollectiveOpHandleBase {
+ public:
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/all_reduce_op_handle.h:40]: Class inheritance is a little complicated.

File: paddle/fluid/framework/details/build_strategy.h [hunk 3]
```diff
@@ -112,6 +112,11 @@ struct BuildStrategy {
   bool cache_expected_kernel_{true};
   std::unordered_set<std::string> mkldnn_enabled_op_types_;
 
+  size_t nccl_comm_num_{1};
+  bool use_hierarchical_allreduce_{false};
+  size_t hierarchical_allreduce_inter_nranks_{0};
+  size_t hierarchical_allreduce_exter_nranks_{0};
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/build_strategy.h:123]: What's the meaning of `hierarchical_allreduce_inter_nranks_ ` and `hierarchical_allreduce_exter_nranks_ `?

File: paddle/fluid/framework/details/fused_all_reduce_op_handle.h [hunk 4]
```diff
@@ -21,21 +21,23 @@
...
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+struct FusedAllReduceOpHandle : public NCCLOpHandleBase {
   FusedAllReduceOpHandle(ir::Node *node,
                          const std::vector<Scope *> &local_scopes,
                          const std::vector<platform::Place> &places,
                          const size_t num_of_all_reduce,
-                         const platform::NCCLContextMap *ctxs);
+                         const platform::MultiNCCLContextMap *ctxs);
 #else
+struct FusedAllReduceOpHandle : public CollectiveOpHandleBase {
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/fused_all_reduce_op_handle.h]: Class inheritance is a little complicated.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:40]: We need to know some ophandles are `CollectiveOpHandleBase `, or we will write code like 
```
if (dynamic_cast<AllReduceOpHandle> != nullptr 
      || dynamic_cast<FusedAllReduceOpHandle> != nullptr
      || ....)
```

File: paddle/fluid/framework/details/build_strategy.h (hunk 3 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/build_strategy.h:123]: image

Comment (gongweibao) [reply paddle/fluid/framework/details/build_strategy.h:123]: Internal(in a node) ranks and external ranks(between nodes).

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:59]: We can polish it in next pr to 
```
#if !defined(USE_NCCL)
#endif
```
Or change CollectiveOpHandleBase to be compatable with CPU and GPU.

File: paddle/fluid/framework/details/fused_all_reduce_op_handle.h (hunk 4 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/fused_all_reduce_op_handle.h]: Same to above.

File: paddle/fluid/framework/details/op_handle_base.cc [hunk 5]
```diff
@@ -187,6 +187,15 @@ void OpHandleBase::RunAndRecordEvent(const std::function<void()> &callback) {
...
     for (auto &p : dev_ctxes_) {
       method = [method, p, this]() {
+        VLOG(10) << "cudadevicecontext:"
+                 << static_cast<platform::CUDADeviceContext *>(p.second)
+                 << ", dev_id:"
+                 << boost::get<platform::CUDAPlace>(p.first).device;
+
+        VLOG(10) << "events:"
+                 << &events_.at(
+                        boost::get<platform::CUDAPlace>(p.first).device);
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/op_handle_base.cc]: Why do you add those code here?  Is it just used to debug for your program?

Comment (gongweibao) [reply paddle/fluid/framework/details/op_handle_base.cc]: Ye. Will clean it!

File: paddle/fluid/platform/nccl_helper.h [hunk 6]
```diff
@@ -124,8 +124,8 @@ struct NCCLContextMap {
           } else {
             rank = trainer_id;
           }
-          VLOG(3) << "init nccl rank: " << rank << " nranks: " << nranks
-                  << " gpu id: " << gpu_id;
+          VLOG(1) << "init nccl rank:" << rank << ", nranks:" << nranks
+                  << ", gpu_id:" << gpu_id << ", dev_id:" << order_[i];
```
Comment (chengduoZH) [inline paddle/fluid/platform/nccl_helper.h:128]: Why change this ?

File: python/paddle/fluid/compiler.py [hunk 7]
```diff
@@ -98,6 +98,7 @@ class CompiledProgram(object):
     def __init__(self, program_or_graph):
         if isinstance(program_or_graph, core.Graph):
             self._graph = program_or_graph
+            # don't not create a new program here.
```
Comment (chengduoZH) [inline python/paddle/fluid/compiler.py:101]: Why?

File: paddle/fluid/platform/nccl_helper.h (hunk 6 above)
Comment (gongweibao) [reply paddle/fluid/platform/nccl_helper.h:128]: It's important info for user to check if the nccl run ok or not!

File: python/paddle/fluid/framework.py [hunk 8]
```diff
@@ -2759,6 +2759,10 @@ def __init__(self):
 
         # use Deep gradient comrepssion or not
         self._enable_dgc = False
+        self._nccl_comm_num = 1
+        self._use_hierarchical_allreduce = False
+        self._hierarchical_allreduce_inter_nranks = 0
+        self._hierarchical_allreduce_exter_nranks = 0
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py:2767]: Why add those attributions here? Those attributions are not related program.

**Comment by chengduoZH** [comment]
This PR should be polished.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:59]: I prefer to fix it in this PR.

File: paddle/fluid/framework/details/build_strategy.h (hunk 3 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/build_strategy.h:123]: Please add doc in code.

File: python/paddle/fluid/framework.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py:2767]: It's needed both in transpiler and build-strategy and we want the use set these attributes once.

Comment (gongweibao) [reply python/paddle/fluid/framework.py:2767]: You can see 301~310 lines.

File: python/paddle/fluid/compiler.py (hunk 7 above)
Comment (gongweibao) [reply python/paddle/fluid/compiler.py:101]: The reason is below.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:59]: Done.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:40]: Done.

File: paddle/fluid/framework/details/build_strategy.h (hunk 3 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/build_strategy.h:123]: Done.

File: paddle/fluid/framework/details/op_handle_base.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/op_handle_base.cc]: Done.

File: paddle/fluid/framework/details/all_reduce_op_handle.h [hunk 9]
```diff
@@ -21,20 +21,23 @@
...
- public:
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+class AllReduceOpHandle : public NCCLOpHandleBase {
+ public:
   AllReduceOpHandle(ir::Node *node, const std::vector<Scope *> &local_scopes,
                     const std::vector<platform::Place> &places,
-                    const platform::NCCLContextMap *ctxs);
+                    const platform::MultiNCCLContextMap *ctxs);
 #else
+class AllReduceOpHandle : public OpHandleBase {
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/all_reduce_op_handle.h:39]: I think that `NCCLOpHandleBase ` is unnecessary.
If I compiled Paddle with GPU, and I want to use it to run a program with DataParallel on CPU. In this case, `AllReduceOpHandle ` inherited from `NCCLOpHandleBase`, it is confusing.

File: paddle/fluid/framework/details/all_reduce_op_handle.h [hunk 10]
```diff
@@ -46,13 +49,17 @@ class AllReduceOpHandle : public OpHandleBase {
...
  protected:
   void RunImpl() override;
-
   std::vector<Scope *> local_scopes_;
+
+#if !(defined(PADDLE_WITH_CUDA) && !defined(_WIN32))
+  // NCCLOpHandleBase already have these attributes.
+  // Will polish it by class inheritance framework.
   std::vector<platform::Place> places_;
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/all_reduce_op_handle.h:58]: If we remove `NCCLOpHandleBase `, the above fix is unnecessary.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  virtual ~NCCLOpHandleBase() {
+    for (auto& ev : inter_events_) {
+      PADDLE_ENFORCE(cudaEventDestroy(ev.second));
+    }
+    for (auto& ev : exter_events_) {
+      PADDLE_ENFORCE(cudaEventDestroy(ev.second));
+    }
+  }
+  void SetRunEnv(int run_order, bool use_hierarchical_allreduce) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h]: The condition is `run_order_ >= 0` but the info is `run_order must > 0`.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    run_order_ = run_order;
+    use_hierarchical_allreduce_ = use_hierarchical_allreduce;
+
+    VLOG(10) << "SetRunEnv "
+             << " run_order:" << run_order
+             << ", use_hierarchical_allreduce:" << use_hierarchical_allreduce;
+
+    if (nccl_ctxs_ == nullptr) {
+      return;
+    }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:66]: This check should be placed in the first line of this function.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+    if (!use_hierarchical_allreduce_) {
+      auto ctxs = nccl_ctxs_->GetFlatCtx(run_order);
+      for (auto& p : places_) {
+        this->SetDeviceContext(p, ctxs->DevCtx(p));
+      }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:72]: If `use_hierarchical_allreduce_ ` is false, don't you need CUDAEvent?

File: paddle/fluid/framework/ir/multi_devices_graph_pass/all_reduce_deps_pass.cc [hunk 14]
```diff
@@ -35,9 +36,20 @@ namespace ir {
...
+    std::vector<details::OpHandleBase*> all_reduce_op_handles =
         GetSortedAllReduceOps(*graph);
 
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+    auto use_hierarchical_allreduce =
+        Get<bool>(details::kUseHierarchicalAllReduce);
+    for (size_t i = 0; i < all_reduce_op_handles.size(); ++i) {
+      auto op_handle =
+          dynamic_cast<details::NCCLOpHandleBase*>(all_reduce_op_handles[i]);
+      PADDLE_ENFORCE(op_handle, "must be nccl_op_handle");
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/multi_devices_graph_pass/all_reduce_deps_pass.cc]: Please complete the error message.

File: paddle/fluid/platform/nccl_helper.h [hunk 15]
```diff
@@ -160,6 +160,151 @@ struct NCCLContextMap {
...
+
+  std::vector<NCCLContextMap *> *GetHierarchicalExterCtxs() {
+    return &h_exter_ctxs_;
+  }
+
+ protected:
+  std::vector<NCCLContextMap *> flat_ctxs_;
+
+  std::vector<NCCLContextMap *> h_inter_ctxs_;
+  std::vector<NCCLContextMap *> h_exter_ctxs_;
```
Comment (chengduoZH) [inline paddle/fluid/platform/nccl_helper.h]: Please explain the difference between those NCCLContextMaps.

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc [hunk 16]
```diff
@@ -118,23 +250,31 @@ GenNCCLId operator
...
+                 "The number of nccl communicator num.")
+        .SetDefault(1);
+    AddAttr<bool>("use_hierarchical_allreduce",
+                  "(bool default false) "
+                  "Wheter to use hierarchical allreduce.")
+        .SetDefault(false);
+    AddAttr<int>("hierarchical_allreduce_inter_nranks",
+                 "(int default 1) "
+                 "Wheter to use hierarchical allreduce.")
+        .SetDefault(-1);
```
Comment (chengduoZH) [inline paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc:272]: Considering compatibility issues, you should add the default value for those added attributions.

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc [hunk 17]
```diff
@@ -118,23 +250,31 @@ GenNCCLId operator
...
   }
 };
 
-}  // namespace operators
-}  // namespace paddle
+};  // namespace operators
+};  // namespace paddle
```
Comment (chengduoZH) [inline paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc]: Please remove `;`.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 9 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:39]: If we don't have ncclhandlebase, the interface it provide move to OphandleBase?
I don't think it's better.
I think there should be a `CollectiveOpHandleBase` compatible both with CPU and GPU and all collective operator handle should inherit from it.
But it should change much larger and it's not better todo in this PR.

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc (hunk 17 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 11 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h]: Done.Thanks!

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 12 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:66]: NO, it's for CPU env with GPU complation.
And these will be polished when we cleanup with `CollectiveOpHandleBase`

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 13 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:72]: Yes.

File: paddle/fluid/framework/ir/multi_devices_graph_pass/all_reduce_deps_pass.cc (hunk 14 above)
Comment (gongweibao) [reply paddle/fluid/framework/ir/multi_devices_graph_pass/all_reduce_deps_pass.cc]: Done.Thanks.

File: paddle/fluid/platform/nccl_helper.h (hunk 15 above)
Comment (gongweibao) [reply paddle/fluid/platform/nccl_helper.h]: Done.Thanks!

File: paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc (hunk 16 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed_ops/gen_nccl_id_op.cc:272]: There should be an invalid value to check it later when running.

File: paddle/fluid/framework/details/all_reduce_op_handle.h (hunk 10 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_op_handle.h:58]: TODO: https://github.com/PaddlePaddle/Paddle/issues/17513

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    for (auto& ev : inter_events_) {
+      PADDLE_ENFORCE(cudaEventDestroy(ev.second));
+    }
+    for (auto& ev : exter_events_) {
+      PADDLE_ENFORCE(cudaEventDestroy(ev.second));
+    }
+  }
+  void SetRunEnv(int run_order, bool use_hierarchical_allreduce) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must >= 0");
+    run_order_ = run_order;
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:57]: ？

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+             << ", stream:" << stream;
+
+    PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
+        sendbuff, recvbuff, count, datatype, ncclSum, comm, stream));
+    if (FLAGS_sync_nccl_allreduce) {
+      PADDLE_ENFORCE(cudaStreamSynchronize(stream),
+                     "sync HierarchicalAllReduce inter stream error");
+    } else {
+      cudaEventRecord(inter_events_.at(dev_id), stream);
+    }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:167]: If `FLAGS_sync_nccl_allreduce` is open, `cudaEventRecord ` should be called also.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+
+  void HierarchicalAllReduce(platform::Place place, const void* sendbuff,
+                             void* recvbuff, size_t count,
+                             ncclDataType_t datatype, ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    InterAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+    if (nccl_ctxs_->NeedExterAllReduce()) {
+      ExterAllReduce(place, recvbuff, recvbuff, count, datatype, op);
+    }
+    InterBroadCast(place, recvbuff, count, datatype, op);
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:141]: Why do you call `InterBroadCast ` here?

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+    auto comm = nccl_ctx.comm_;
+
+    VLOG(10) << "before all reduce run_order:" << run_order_
+             << "buffer:" << sendbuff << ", numel:" << count
+             << ", dev_id:" << dev_id << ", dtype:" << datatype
+             << ", place:" << place << ", stream:" << stream;
+
+    if (!FLAGS_sync_nccl_allreduce) {
+      cudaStreamWaitEvent(stream, inter_events_.at(dev_id), 0);
+    }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h]: `FLAGS_sync_nccl_allreduce` should not be used like this.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+    cudaStreamWaitEvent(stream, exter_events_.at(dev_id), 0);
+    PADDLE_ENFORCE(platform::dynload::ncclBcast(sendbuff, count, datatype, 0,
+                                                comm, stream));
+  }
+
+ protected:
+  std::vector<platform::Place> places_;
+  const platform::MultiNCCLContextMap* nccl_ctxs_{nullptr};
+  int run_order_{0};
+  bool use_hierarchical_allreduce_{false};
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:224]: Please add doc for `run_order_ ` and `use_hierarchical_allreduce_ `.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 20 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/nccl_op_handle.h:141]: I see.

Comment (chengduoZH) [reply paddle/fluid/framework/details/nccl_op_handle.h:141]: If the `nccl_ctxs_->NeedExterAllReduce()` if false, not need to call `InterBroadCast `.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    HierarchicalAllReduce(place, sendbuff, recvbuff, count, datatype, op);
+  }
+
+  void HierarchicalAllReduce(platform::Place place, const void* sendbuff,
+                             void* recvbuff, size_t count,
+                             ncclDataType_t datatype, ncclRedOp_t op) {
+    PADDLE_ENFORCE(run_order_ >= 0, "run_order must > 0");
+    InterAllReduce(place, sendbuff, recvbuff, count, datatype, op);
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h]: There should use `Reduce`, and it will be a bit faster than `AllReduce`.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 18 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:57]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 20 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:141]: We need.
Some trainer need not exter allreduce because they are not in exter allreduce ring, but when broadcast they need call broadcast all.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 21 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 22 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:224]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 23 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h]: Done.Thanks!

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 20 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/nccl_op_handle.h:141]: I think it will be better:
```
if (nccl_ctxs_->NeedExterAllReduce()) {
 InterReduce(place, sendbuff, recvbuff, count, datatype, op);
 ExterAllReduce(place, recvbuff, recvbuff, count, datatype, op);
 InterBroadCast(place, recvbuff, count, datatype, op);
} else {
 FlatNcclAllReduce(place, sendbuff, recvbuff, count, datatype, op);
}
```
Because InterReduce and InterBroadCast have some stream sync operation, maybe it will affect the speed of origin implementation.

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+    for (auto& p : places_) {
+      auto ctxs = nccl_ctxs_->GetHierarchicalInterCtx(run_order);
+      this->SetDeviceContext(p, ctxs->DevCtx(p));
+    }
+
+    for (auto& p : dev_ctxes_) {
+      int dev_id = boost::get<platform::CUDAPlace>(p.first).device;
+      if (inter_events_.find(dev_id) != inter_events_.end()) {
+        continue;
+      }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h:88]: Why do you write line 86 ~88?

File: paddle/fluid/framework/details/nccl_op_handle.h [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+                                              cudaEventDisableTiming));
+      PADDLE_ENFORCE(cudaEventCreateWithFlags(&exter_events_[dev_id],
+                                              cudaEventDisableTiming));
+      VLOG(10) << "Create events on dev_id:" << dev_id
+               << ", inter_event:" << &inter_events_[dev_id]
+               << ", exter_event:" << &exter_events_[dev_id];
+    }
+  }
+
+  void FlatNcclAllReduce(platform::Place place, const void* sendbuff,
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/nccl_op_handle.h]: `FlatNcclAllReduce ` -> `FlatNCCLAllReduce `

File: paddle/fluid/platform/nccl_helper.h [hunk 26]
```diff
@@ -160,6 +160,154 @@ struct NCCLContextMap {
...
+  snprintf(name, sizeof(name) - 1, "%s_%d", NCCL_ID_VARNAME,
+           static_cast<int>(pos));
+  return name;
+}
+
+inline std::string GetHierarchicalExterNCCLVarName(size_t pos) {
+  char name[256];
+  snprintf(name, sizeof(name) - 1, "Hierarchical_exter_%s_%d", NCCL_ID_VARNAME,
+           static_cast<int>(pos));
+  return name;
```
Comment (chengduoZH) [inline paddle/fluid/platform/nccl_helper.h]: =>
```
return string::Sprintf("Hierarchical_exter_%s_%d", NCCL_ID_VARNAME,
           static_cast<int>(pos)) ;
```

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 27]
```diff
@@ -160,6 +160,15 @@ class DistributeTranspilerConfig(object):
     runtime_split_send_recv = False
     sync_mode = None
 
+    nccl_comm_num = 1
+    #The picture is here:
```
Comment (chengduoZH) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: ?

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 19 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:167]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 20 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:141]: You missed it. Talk about it later.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 25 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h]: Done.

File: paddle/fluid/platform/nccl_helper.h (hunk 26 above)
Comment (gongweibao) [reply paddle/fluid/platform/nccl_helper.h]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 27 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: paddle/fluid/framework/details/nccl_op_handle.h (hunk 24 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/nccl_op_handle.h:88]: If the event hasn't been created then create it!

File: paddle/fluid/platform/nccl_helper.h [hunk 28]
```diff
@@ -160,6 +160,147 @@ struct NCCLContextMap {
...
+    return &h_exter_ctxs_;
+  }
+
+ protected:
+  // Support multi nccl comm on default nccl ring while NCCLContextMap can't.
+  std::vector<NCCLContextMap *> flat_ctxs_;
+
+  // h_inter_ctxs_ and h_exter_ctxs_ are for 2d allreduce.
+  // And h_exter_ctxs_ can support multi comm too.
+  std::vector<NCCLContextMap *> h_inter_ctxs_;
```
Comment (chengduoZH) [inline paddle/fluid/platform/nccl_helper.h]: Why not use std::unique_ptr here?

File: paddle/fluid/platform/nccl_helper.h [hunk 29]
```diff
@@ -160,6 +160,147 @@ struct NCCLContextMap {
...
+
+    PADDLE_ENFORCE(inter_trainers_num > 1, "inter_trainers_num:%llu must > 1",
+                   inter_trainers_num);
+
+    int inter_trainer_id = trainer_id % inter_trainers_num;
+    VLOG(1) << "init inter_trainer_id:" << inter_trainer_id;
+    auto local = new NCCLContextMap(places, inter_nccl_id, inter_trainers_num,
+                                    inter_trainer_id);
+
+    h_inter_ctxs_.push_back(local);
```
Comment (chengduoZH) [inline paddle/fluid/platform/nccl_helper.h]: I prefer to use std::unique_ptr here.

File: paddle/fluid/platform/nccl_helper.h (hunk 28 above)
Comment (gongweibao) [reply paddle/fluid/platform/nccl_helper.h]: Free them in destructor function.

File: paddle/fluid/platform/nccl_helper.h (hunk 29 above)
Comment (gongweibao) [reply paddle/fluid/platform/nccl_helper.h]: Free them in destructor function and the vector destructs itself.

File: python/paddle/fluid/framework.py (hunk 8 above)
Comment (chengduoZH) [reply python/paddle/fluid/framework.py:2767]: I think it is not appropriate that add those options in Program. It should be fixed in the next PR.

**Review by chengduoZH — APPROVED** [review]

File: python/paddle/fluid/framework.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/fluid/framework.py:2767]: https://github.com/PaddlePaddle/Paddle/issues/17617

File: paddle/fluid/framework/details/build_strategy.h (hunk 3 above)
Comment (guru4elephant) [reply paddle/fluid/framework/details/build_strategy.h:123]: You should cite the reference of original image. https://arxiv.org/pdf/1807.11205.pdf
